File: morphac/simulation/playground/test/playground_test.cc
#include "Eigen/Dense"
#include "gtest/gtest.h"

#include "mechanics/models/include/diffdrive_model.h"
#include "simulation/playground/include/playground.h"

namespace {

using std::make_unique;
using std::srand;
using std::sqrt;
using std::unique_ptr;

using Eigen::MatrixXd;
using Eigen::VectorXd;

using morphac::constructs::ControlInput;
using morphac::constructs::State;
using morphac::environment::Map;
using morphac::mechanics::models::DiffdriveModel;
using morphac::math::numeric::IntegratorType;
using morphac::robot::blueprint::Footprint;
using morphac::robot::blueprint::Robot;
using morphac::robot::pilot::Pilot;
using morphac::simulation::playground::PlaygroundSpec;
using morphac::simulation::playground::PlaygroundState;
using morphac::simulation::playground::Playground;

// Global kinematic models as if they're destroyed, the reference to the model
// in Robot points to nothing and we get strange results.
DiffdriveModel diffdrive_model(1., 1.);

// Derived class from Pilot for testing.
class CustomPilot : public Pilot {
 public:
  CustomPilot(VectorXd control_input_data)
      : Pilot(), control_input_data_(control_input_data) {}

  ControlInput Execute(const PlaygroundState& playground_state,
                       const int uid) const override {
    // Some asserts to prevent the unused variable warning.
    MORPH_REQUIRE(playground_state.NumRobots() >= 0, std::invalid_argument,
                  "Invalid PlaygroundState");
    MORPH_REQUIRE(uid >= 0, std::invalid_argument, "UID must be non negative");
    return ControlInput(control_input_data_);
  }

 private:
  VectorXd control_input_data_;
};

class PlaygroundTest : public ::testing::Test {
 protected:
  PlaygroundTest() {
    // Set random seed for Eigen.
    srand(7);

    robot1_ =
        make_unique<Robot>(diffdrive_model, Footprint(MatrixXd::Zero(20, 2)));
    robot2_ =
        make_unique<Robot>(diffdrive_model, Footprint(MatrixXd::Zero(20, 2)),
                           State({1., 2., 0.}, {}));
    robot3_ =
        make_unique<Robot>(diffdrive_model, Footprint(MatrixXd::Zero(20, 2)),
                           State({-5., 7., 1.57}, {}));
    playground_ =
        make_unique<Playground>(playground_spec_, Map(map_data_, 0.1));
  }

  void SetUp() override {}

  unique_ptr<Playground> playground_;
  unique_ptr<Robot> robot1_, robot2_, robot3_;
  MatrixXd map_data_ = MatrixXd::Random(20, 10);
  PlaygroundSpec playground_spec_{"playground", 0.05};
};

TEST_F(PlaygroundTest, GetState) {
  // Making sure that we get can get the correct state from the playground.
  PlaygroundState& playground_state = playground_->get_state();

  ASSERT_EQ(playground_state.get_time(), 0);

  ASSERT_TRUE(playground_state.get_map().get_data().isApprox(map_data_));
  ASSERT_EQ(playground_state.NumRobots(), 0);
}

TEST_F(PlaygroundTest, SetState) {
  // Making sure that the state reference works correctly. If we change the
  // value of the state through the reference, it should reflect back in the
  // state stored by the Playground.
  PlaygroundState& playground_state = playground_->get_state();

  playground_state.set_time(0.002);
  ASSERT_EQ(playground_->get_state().get_time(), 0.002);

  MatrixXd new_map_data = MatrixXd::Random(20, 10);
  Map new_map{new_map_data, 0.1};
  playground_state.set_map(new_map);

  ASSERT_TRUE(
      playground_->get_state().get_map().get_data().isApprox(new_map_data));

  // Testing adding robots.
  // We only check the number of robots in the state. Addition of the correct
  // robots against the uids is checked by the PlaygroundState test.
  playground_state.AddRobot(*robot1_, 0);
  ASSERT_EQ(playground_->get_state().NumRobots(), 1);

  // Adding the other two robots.
  playground_state.AddRobot(*robot2_, 1);
  playground_state.AddRobot(*robot3_, 2);
  ASSERT_EQ(playground_->get_state().NumRobots(), 3);
}

TEST_F(PlaygroundTest, AddRobot) {
  // Add the first robot.
  playground_->AddRobot(*robot1_, CustomPilot(VectorXd::Zero(3)),
                        IntegratorType::kEulerIntegrator, 1);

  // Test if the robot has been added. As this is playground's AddRobot
  // interface, we also test if the correct robot has been added by checking
  // the initial state
  ASSERT_EQ(playground_->get_state().NumRobots(), 1);
  ASSERT_TRUE(
      playground_->get_state().get_robot(1).get_state().get_data().isApprox(
          VectorXd::Zero(3)));

  // Adding the other robots.
  // Initial state vectors for the other robots.
  VectorXd initial_state_data1(3), initial_state_data2(3);
  initial_state_data1 << 1., 2., 0.;
  initial_state_data2 << -5., 7., 1.57;

  playground_->AddRobot(*robot2_, CustomPilot(VectorXd::Ones(3)),
                        IntegratorType::kMidPointIntegrator, 2);
  playground_->AddRobot(*robot3_, CustomPilot(VectorXd::Random(3)),
                        IntegratorType::kRK4Integrator, 3);

  ASSERT_EQ(playground_->get_state().NumRobots(), 3);
  ASSERT_TRUE(
      playground_->get_state().get_robot(2).get_state().get_data().isApprox(
          initial_state_data1));
  ASSERT_TRUE(
      playground_->get_state().get_robot(3).get_state().get_data().isApprox(
          initial_state_data2));
}

TEST_F(PlaygroundTest, InvalidAddRobot) {
  // Add the first robot.
  playground_->AddRobot(*robot1_, CustomPilot(VectorXd::Zero(3)),
                        IntegratorType::kEulerIntegrator, 7);

  // Add the other robots with invalid uids.
  // Negative uid.
  ASSERT_THROW(playground_->AddRobot(*robot2_, CustomPilot(VectorXd::Zero(3)),
                                     IntegratorType::kMidPointIntegrator, -1),
               std::logic_error);
  // Duplicate uid.
  ASSERT_THROW(playground_->AddRobot(*robot3_, CustomPilot(VectorXd::Zero(3)),
                                     IntegratorType::kRK4Integrator, 7),
               std::logic_error);
}

}  // namespace

int main(int argc, char** argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: morphac/simulation/playground/python/test/playground_test.py
import numpy as np
import pytest

from morphac.constructs import ControlInput, State
from morphac.environment import Map
from morphac.mechanics.models import DiffdriveModel
from morphac.math.numeric import (
    IntegratorType,
    Integrator,
    MidPointIntegrator,
    RK4Integrator,
)
from morphac.robot.pilot import Pilot
from morphac.robot.blueprint import Footprint, Robot
from morphac.simulation.playground import Playground
from morphac.simulation.playground import PlaygroundSpec


# Custom derived Pilot class to be used for testing.
class CustomPilot(Pilot):
    def __init__(self, control_input_data):

        Pilot.__init__(self)
        self.control_input_data = control_input_data

    def execute(self, playground_state, uid):

        return ControlInput(self.control_input_data)


@pytest.fixture()
def generate_playground():
    playground_spec = PlaygroundSpec("playground", 0.01)
    playground = Playground(playground_spec, Map(50, 50, 0.1))

    return playground


@pytest.fixture()
def generate_robot_list():
    r1 = Robot(DiffdriveModel(1.0, 1.0), Footprint([[0, 0]]))
    r2 = Robot(
        DiffdriveModel(1.0, 1.0), Footprint([[0, 0]]), State([1.0, 2.0, 0.0], [])
    )

    return r1, r2


def test_spec(generate_playground):
    playground = generate_playground

    # Testing the specs of the playground.
    assert playground.spec.name == "playground"
    assert playground.spec.dt == 0.01


def test_state_read(generate_playground):
    # Make sure that the correct playground state can be obtained from the
    # playground
    playground = generate_playground
    playground_state = playground.state

    # Making sure that the map is correct.
    assert np.allclose(playground_state.map.data, np.zeros([500, 500]))

    # Making sure that there are currently no robots.
    assert playground_state.num_robots == 0

    # Testing default time.
    assert playground_state.time == 0


def test_state_write(generate_playground, generate_robot_list):
    # Make sure that the returned state is a reference and the value of the
    # actual internal state can be changed through this state.
    playground = generate_playground
    robot1, robot2 = generate_robot_list

    playground_state = playground.state

    # Change time.
    playground_state.time = 0.01

    assert playground.state.time == 0.01

    playground_state.map = Map(np.ones([500, 500]), 0.1)

    # The state value in the Playground object should have changed.
    assert np.allclose(playground.state.map.data, np.ones([500, 500]))

    # Adding a robot through the state. Note that this should never be carried
    # out unless its for testing purposes. Playground's add_robot function
    # must always be used.
    playground_state.add_robot(robot1, 0)
    playground_state.add_robot(robot2, 1)

    # Make sure that the state object and internal playground state both have
    # two robots.
    assert playground_state.num_robots == 2
    assert playground.state.num_robots == 2


def test_get_pilot_oracle(generate_playground, generate_robot_list):
    playground = generate_playground
    robot1, robot2 = generate_robot_list

    pilot1 = CustomPilot([0, 0])
    pilot2 = CustomPilot([1.0, 2])

    # Adding the robots.
    playground.add_robot(robot1, pilot1, IntegratorType.EULER_INTEGRATOR, 1)
    playground.add_robot(robot2, pilot2, IntegratorType.MID_POINT_INTEGRATOR, 2)

    pilot_oracle = playground.get_pilot_oracle()

    # Making sure the oracle is a dict object as is wrapped from an unordered
    # map.
    assert isinstance(pilot_oracle, dict)

    # Length of the dict.
    assert len(pilot_oracle) == 2

    # Individual pilot elements.
    assert np.allclose(pilot_oracle[1].control_input_data, [0, 0])
    assert np.allclose(pilot_oracle[2].control_input_data, [1, 2])

    # Invalid oracle key.
    with pytest.raises(KeyError):
        _ = pilot_oracle[0]
    with pytest.raises(KeyError):
        _ = pilot_oracle[3]


def test_get_pilot(generate_playground, generate_robot_list):
    playground = generate_playground
    robot1, robot2 = generate_robot_list

    pilot1 = CustomPilot([0, 0])
    pilot2 = CustomPilot([1.0, 2])

    # Adding the robots.
    playground.add_robot(robot1, pilot1, IntegratorType.EULER_INTEGRATOR, 1)
    playground.add_robot(robot2, pilot2, IntegratorType.MID_POINT_INTEGRATOR, 2)

    # Making sure the right pilots are returned.
    assert np.allclose(playground.get_pilot(1).control_input_data, [0, 0])
    assert np.allclose(playground.get_pilot(2).control_input_data, [1.0, 2])

    # Invalid uid.
    with pytest.raises(ValueError):
        _ = playground.get_pilot(0)
    with pytest.raises(ValueError):
        _ = playground.get_pilot(3)


def test_get_integrator(generate_playground, generate_robot_list):
    playground = generate_playground
    robot1, robot2 = generate_robot_list

    pilot1 = CustomPilot([0, 0])
    pilot2 = CustomPilot([1.0, 2])

    # Adding the robots.
    playground.add_robot(robot1, pilot1, IntegratorType.MID_POINT_INTEGRATOR, 1)
    playground.add_robot(robot2, pilot2, IntegratorType.RK4_INTEGRATOR, 2)

    # Making sure the right integrators are returned.
    assert isinstance(playground.get_integrator(1), MidPointIntegrator)
    assert isinstance(playground.get_integrator(2), RK4Integrator)

    # Invalid uid.
    with pytest.raises(ValueError):
        _ = playground.get_integrator(0)
    with pytest.raises(ValueError):
        _ = playground.get_integrator(3)


def test_add_robot(generate_playground, generate_robot_list):
    playground = generate_playground
    robot1, robot2 = generate_robot_list

    pilot1 = CustomPilot([0, 0])
    pilot2 = CustomPilot([1.0, 2])

    playground.add_robot(robot1, pilot1, IntegratorType.EULER_INTEGRATOR, 1)
    playground.add_robot(robot2, pilot2, IntegratorType.MID_POINT_INTEGRATOR, 2)

    # Make sure that the robots have been added correctly.
    assert playground.state.num_robots == 2

    assert np.allclose(playground.state.get_robot_state(1).data, robot1.state.data)
    assert np.allclose(playground.state.get_robot_state(2).data, robot2.state.data)


def test_add_robot_with_temporary(generate_playground):
    # Test the function with temporarily created Robot and Pilot objects.
    # This is to test for any cpp lifetime management weirdness.

    playground = generate_playground

    playground.add_robot(
        Robot(DiffdriveModel(1.0, 1.0)),
        CustomPilot([0, 0]),
        IntegratorType.EULER_INTEGRATOR,
        1,
    )
    playground.add_robot(
        Robot(DiffdriveModel(1.0, 1.0)),
        CustomPilot([1, 2]),
        IntegratorType.MID_POINT_INTEGRATOR,
        2,
    )

    # Make sure that the robots have been added correctly.
    assert playground.state.num_robots == 2

    assert np.allclose(playground.state.get_robot_state(1).data, [0.0, 0.0, 0.0])
    assert np.allclose(playground.state.get_robot_state(2).data, [0.0, 0.0, 0.0])


def test_execute(generate_playground, generate_robot_list):
    playground = generate_playground
    robot1, robot2 = generate_robot_list
    pilot1, pilot2 = CustomPilot([0, 0]), CustomPilot([1.0, 1.0])

    # Adding the robots.
    playground.add_robot(robot1, pilot1, IntegratorType.MID_POINT_INTEGRATOR, 1)
    playground.add_robot(robot2, pilot2, IntegratorType.RK4_INTEGRATOR, 2)

    # Make sure that the playground time is 0.
    assert playground.state.time == 0.0

    # Executing a playground cycle.
    playground.execute()

    # Test the states of the robots.
    # The first pilot gives zero control inputs, hence the state must remain
    # unchanged.
    assert np.allclose(playground.state.get_robot_state(1).data, [0.0, 0.0, 0.0])

    # The second robot receives a constant input of [1., 1.]. As the wheel
    # radius = 1 m, the robot moves forward with a velocity of 1 m/s. Hence
    # it moves a total distance of dt m in the x direction (The initial angle
    # is zero).
    assert np.allclose(
        playground.state.get_robot_state(2).data, [1.0 + playground.spec.dt, 2.0, 0.0]
    )

    # Make sure that the playground time has updated.
    assert playground.state.time == playground.spec.dt


def test_execute_with_temporary(generate_playground):
    # Test the function after adding robots using temporarily created Robot and Pilot objects.
    # This is to test for any cpp lifetime management weirdness.

    playground = generate_playground

    playground.add_robot(
        Robot(DiffdriveModel(1.0, 1.0)),
        CustomPilot([0, 0]),
        IntegratorType.EULER_INTEGRATOR,
        1,
    )
    playground.add_robot(
        Robot(DiffdriveModel(1.0, 1.0), initial_state=State([1.0, 2.0, 0.0], [])),
        CustomPilot([1, 1]),
        IntegratorType.MID_POINT_INTEGRATOR,
        2,
    )

    # Make sure that the playground time is 0.
    assert playground.state.time == 0.0

    # Executing a playground cycle.
    playground.execute()

    # Test the states of the robots.
    # The first pilot gives zero control inputs, hence the state must remain
    # unchanged.
    assert np.allclose(playground.state.get_robot_state(1).data, [0.0, 0.0, 0.0])

    # The second robot receives a constant input of [1., 1.]. As the wheel
    # radius = 1 m, the robot moves forward with a velocity of 1 m/s. Hence
    # it moves a total distance of dt m in the x direction (The initial angle
    # is zero).
    assert np.allclose(
        playground.state.get_robot_state(2).data, [1.0 + playground.spec.dt, 2.0, 0.0]
    )

    # Make sure that the playground time has updated.
    assert playground.state.time == playground.spec.dt

